File: src/org/usfirst/frc/team4099/SmartDashboardInteractions.java
package org.usfirst.frc.team4099;

import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import org.usfirst.frc.team4099.auto.AutoModeEndedException;
import org.usfirst.frc.team4099.auto.modes.*;
import org.usfirst.frc.team4099.lib.util.AutonomousInitParameters;
import org.usfirst.frc.team4099.lib.util.Rotation2D;

import java.util.ArrayList;

/**
 * Controls the interactive elements of SmartDashboard.
 *
 * Keeps the network tables keys in one spot and enforces autonomous mode
 * invariants.
 */
public class SmartDashboardInteractions {

    private static final String HOOD_TUNING_MODE = "Hood Tuning Mode";
    private static final String OUTPUT_TO_SMART_DASHBOARD = "Output To SmartDashboard";

    private static final String AUTO_OPTIONS = "auto_options";
    private static final String SELECTED_AUTO_MODE = "selected_auto_mode";
    private static final String SELECTED_AUTO_LANE = "selected_auto_lane";

    private static final AutonOption DEFAULT_MODE = AutonOption.ONE_GEAR;
    private static final AutonLane DEFAULT_LANE = AutonLane.LEFT_LANE;

    public void initWithDefaults() {
        SmartDashboard.putBoolean(HOOD_TUNING_MODE, false);
        SmartDashboard.putBoolean(OUTPUT_TO_SMART_DASHBOARD, true);

        ArrayList<String> autoOptionsArray = new ArrayList<>(AutonOption.values().length);
        for (AutonOption autonOption : AutonOption.values()) {
            autoOptionsArray.add(autonOption.name);
        }
        SmartDashboard.putString(AUTO_OPTIONS, autoOptionsArray.toString());
        SmartDashboard.putString(SELECTED_AUTO_MODE, DEFAULT_MODE.name);
        SmartDashboard.putString(SELECTED_AUTO_LANE, DEFAULT_LANE.numberString);
    }

    public boolean isInHoodTuningMode() {
        return SmartDashboard.getBoolean(HOOD_TUNING_MODE, false);
    }

    public boolean shouldLogToSmartDashboard() {
        return SmartDashboard.getBoolean(OUTPUT_TO_SMART_DASHBOARD, true);
    }

    public AutoModeBase getSelectedAutonMode() {
        String autoModeString = SmartDashboard.getString(SELECTED_AUTO_MODE, DEFAULT_MODE.name);
        AutonOption selectedOption = DEFAULT_MODE;
        for (AutonOption autonOption : AutonOption.values()) {
            if (autonOption.name.equals(autoModeString)) {
                selectedOption = autonOption;
                break;
            }
        }

        String autoLaneString = SmartDashboard.getString(SELECTED_AUTO_LANE, DEFAULT_LANE.numberString);
        AutonLane selectedLane = DEFAULT_LANE;
        for (AutonLane autonLane : AutonLane.values()) {
            if (autonLane.numberString.equals(autoLaneString)) {
                selectedLane = autonLane;
            }
        }

        return createAutoMode(selectedOption, selectedLane);
    }


    /**
     * I don't trust SendableChooser to manage {@link AutoModeBase}
     * objects directly, so use this enum to project us from WPILIb.
     */
    enum AutonOption {
        ONE_GEAR("One Gear"),
        ONE_GEAR_AND_BACK_OUT("One Gear, Back Out"),
        ONE_GEAR_AND_BACK_OUT_AND_TURN("One Gear, Back Out, Turn Around"),
        TWO_GEAR("Two Gears"),
        TWO_GEAR_AND_BACK_OUT("Two Gears, Back Out"),
        TWO_GEAR_AND_BACK_OUT_AND_TURN("Two Gears, Back Out, Turn Around"),
        BASELINE("Baseline"),
        BASELINE_AND_TURN("Baseline, Turn Around"),
        STAND_STILL("Stand Still"),
        TEST_DRIVE("TEST ONLY Driving");

        public final String name;

        AutonOption(String name) {
            this.name = name;
        }
    }

    enum AutonLane {
        LEFT_LANE(60, "1"), CENTER_LANE(60, "2"), RIGHT_LANE(75, "3");

        public final double distanceToDrive;
        public final String numberString;

        AutonLane(double distanceToDrive, String numberString) {
            this.distanceToDrive = distanceToDrive;
            this.numberString = numberString;
        }
    }

    private AutonomousInitParameters getAimingHintForLane(AutonLane lane) {
        if (lane == AutonLane.LEFT_LANE) {
            return new AutonomousInitParameters(lane.distanceToDrive, Rotation2D.fromDegrees(60), -1);
        } else if (lane == AutonLane.CENTER_LANE) {
            return new AutonomousInitParameters(lane.distanceToDrive, Rotation2D.fromDegrees(0), -1);
        } else {/* if (lane == AutonLane.RIGHT_LANE) {*/
            return new AutonomousInitParameters(lane.distanceToDrive, Rotation2D.fromDegrees(-60), -1);
        }
    }

    private AutoModeBase createAutoMode(AutonOption autonOption, AutonLane autonLane) {
        switch (autonOption) {
            case ONE_GEAR:
                return new OneGearMode(getAimingHintForLane(autonLane), false, false);
            case ONE_GEAR_AND_BACK_OUT:
                return new OneGearMode(getAimingHintForLane(autonLane), true, false);
            case ONE_GEAR_AND_BACK_OUT_AND_TURN:
                return new OneGearMode(getAimingHintForLane(autonLane), true, true);
            case TWO_GEAR:
                return new TwoGearMode(getAimingHintForLane(autonLane), false, false);
            case TWO_GEAR_AND_BACK_OUT:
                return new TwoGearMode(getAimingHintForLane(autonLane), true, false);
            case TWO_GEAR_AND_BACK_OUT_AND_TURN:
                return new TwoGearMode(getAimingHintForLane(autonLane), true, true);
            case BASELINE:
                return new BaselineMode(getAimingHintForLane(autonLane), false);
            case BASELINE_AND_TURN:
                return new BaselineMode(getAimingHintForLane(autonLane), true);
            case TEST_DRIVE:
                return new AutoModeBase() {
                    @Override
                    protected void routine() throws AutoModeEndedException {
                        throw new RuntimeException("Expected exception!!!");
                    }
                };

            case STAND_STILL: // fallthrough
            default:
                System.out.println("ERROR: unexpected auto mode: " + autonOption);
                return new StandStillMode();
        }
    }
}

File: src/main/java/org/usfirst/frc/team4099/SmartDashboardInteractions.kt
package org.usfirst.frc.team4099

import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import org.usfirst.frc.team4099.auto.AutoModeEndedException
import org.usfirst.frc.team4099.auto.modes.AutoModeBase
import org.usfirst.frc.team4099.auto.modes.BaselineMode
import org.usfirst.frc.team4099.auto.modes.OneGearMode
import org.usfirst.frc.team4099.auto.modes.StandStillMode
import org.usfirst.frc.team4099.lib.util.AutonomousInitParameters
import org.usfirst.frc.team4099.lib.util.Rotation2D
import java.util.*

/**
 * Controls the interactive elements of SmartDashboard.
 *
 * Keeps the network tables keys in one spot and enforces autonomous mode
 * invariants.
 */
class SmartDashboardInteractions {

    fun initWithDefaults() {
        SmartDashboard.putBoolean(HOOD_TUNING_MODE, false)
        SmartDashboard.putBoolean(OUTPUT_TO_SMART_DASHBOARD, true)

        val autoOptionsArray = ArrayList<String>(AutonOption.values().size)
        AutonOption.values().mapTo(autoOptionsArray) { it.nameStr }
        SmartDashboard.putString(AUTO_OPTIONS, autoOptionsArray.toString())
        SmartDashboard.putString(SELECTED_AUTO_MODE, DEFAULT_MODE.name)
        SmartDashboard.putString(SELECTED_AUTO_LANE, DEFAULT_LANE.numberString)
    }

    val isInHoodTuningMode: Boolean
        get() = SmartDashboard.getBoolean(HOOD_TUNING_MODE, false)

    fun shouldLogToSmartDashboard(): Boolean {
        return SmartDashboard.getBoolean(OUTPUT_TO_SMART_DASHBOARD, true)
    }

    val selectedAutonMode: AutoModeBase
        get() {
            val autoModeString = SmartDashboard.getString(SELECTED_AUTO_MODE, DEFAULT_MODE.name)
            var selectedOption = DEFAULT_MODE
            for (autonOption in AutonOption.values()) {
                if (autonOption.nameStr == autoModeString) {
                    selectedOption = autonOption
                    break
                }
            }

            val autoLaneString = SmartDashboard.getString(SELECTED_AUTO_LANE, DEFAULT_LANE.numberString)
            var selectedLane = DEFAULT_LANE
            for (autonLane in AutonLane.values()) {
                if (autonLane.numberString == autoLaneString) {
                    selectedLane = autonLane
                }
            }

            return createAutoMode(selectedOption, selectedLane)
        }


    /**
     * I don't trust SendableChooser to manage [AutoModeBase]
     * objects directly, so use this enum to project us from WPILIb.
     */
    internal enum class AutonOption constructor(val nameStr: String) {
        ONE_GEAR("One Gear"),
        BASELINE("Baseline"),
        TEST_DRIVE("TEST ONLY Driving")
    }

    internal enum class AutonLane private constructor(val distanceToDrive: Double, val numberString: String) {
        LEFT_LANE(2.5, "1"), CENTER_LANE(6.0, "2"), RIGHT_LANE(2.25, "3")
    }

    private fun getAimingHintForLane(lane: AutonLane): AutonomousInitParameters {
        return when (lane) {
            AutonLane.LEFT_LANE -> AutonomousInitParameters(lane.distanceToDrive, Rotation2D.fromDegrees(60.0), -1)
            AutonLane.CENTER_LANE -> AutonomousInitParameters(lane.distanceToDrive, Rotation2D.fromDegrees(0.0), -1)
            else -> /* if (lane == AutonLane.RIGHT_LANE) {*/
                AutonomousInitParameters(lane.distanceToDrive, Rotation2D.fromDegrees(-60.0), -1)
        }
    }

    private fun createAutoMode(autonOption: AutonOption, autonLane: AutonLane): AutoModeBase {
        when (autonOption) {
            SmartDashboardInteractions.AutonOption.ONE_GEAR -> return OneGearMode(getAimingHintForLane(autonLane), false, false)
            SmartDashboardInteractions.AutonOption.BASELINE -> return BaselineMode(getAimingHintForLane(autonLane), false)
            SmartDashboardInteractions.AutonOption.TEST_DRIVE -> return object : AutoModeBase() {
                @Throws(AutoModeEndedException::class)
                override fun routine() {
                    throw RuntimeException("Expected exception!!!")
                }
            }
            else -> {
                println("ERROR: unexpected auto mode: " + autonOption)
                return StandStillMode()
            }
        }
    }

    companion object {

        private val HOOD_TUNING_MODE = "Hood Tuning Mode"
        private val OUTPUT_TO_SMART_DASHBOARD = "Output To SmartDashboard"

        private val AUTO_OPTIONS = "auto_options"
        private val SELECTED_AUTO_MODE = "selected_auto_mode"
        private val SELECTED_AUTO_LANE = "selected_auto_lane"

        private val DEFAULT_MODE = AutonOption.ONE_GEAR
        private val DEFAULT_LANE = AutonLane.CENTER_LANE
    }
}